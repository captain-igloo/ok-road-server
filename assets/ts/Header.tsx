import * as React from 'react';

import UserMenu from './UserMenu';

interface Props {
    showDemo?: boolean;
    showMap?: boolean;
    showRegister?: boolean;
    showSignIn?: boolean;
    user?: {
        fullName: string;
    };
}

export default function Header(props: Props) {
    const {
        showDemo,
        showMap,
        showRegister,
        showSignIn,
        user,
    } = props;

    return (
        <div className="header">
            <div className="logo">
                <a href="/">
                    <img alt="Home" src="/img/ok.svg" width="40px" height="40px" />
                </a>
                &nbsp;
                OK Road New Zealand
            </div>
            <UserMenu
                showDemo={showDemo}
                showMap={showMap}
                showRegister={showRegister}
                showSignIn={showSignIn}
                user={user}
            />
        </div>
    );
}
