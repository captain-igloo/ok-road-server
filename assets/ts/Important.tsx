import * as React from 'react';

interface Props {
    children: React.ReactNode;
}

export default function Important(props: Props) {
    const { children } = props;

    return (
        <div className="alert__important">
            <p className="alert__important--title">
                <img alt="" src="/img/important.svg" width={16} />
                {' '}
                Important
            </p>
            <p className="alert__important--body">
                {children}
            </p>
        </div>
    );
}
