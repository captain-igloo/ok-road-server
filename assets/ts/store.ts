import { combineReducers, configureStore } from '@reduxjs/toolkit';

import configReducer from './config/slice';
import friendsReducer from './friends/slice';
import mapReducer from './map/slice';
import notificationsReducer from './notifications/slice';

const rootReducer = combineReducers({
    config: configReducer,
    friends: friendsReducer,
    notifications: notificationsReducer,
    map: mapReducer,
});

export const setupStore = (preloadedState?: Partial<RootState>) => configureStore({
    reducer: rootReducer,
    preloadedState,
});

export type AppStore = ReturnType<typeof setupStore>;

export type RootState = ReturnType<typeof rootReducer>;

export type AppDispatch = AppStore['dispatch'];
