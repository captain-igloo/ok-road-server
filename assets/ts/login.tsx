import * as React from 'react';
import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';
import Card from 'react-bootstrap/Card';
import Form from 'react-bootstrap/Form';
import * as ReactDOM from 'react-dom/client';

import Header from './Header';

declare global {
    interface Window {
        error: string;
        token: string;
    }
}

const element = document.getElementById('login');

if (element) {
    const root = ReactDOM.createRoot(element);

    let alert;
    if (window.error) {
        alert = <Alert variant="danger">{window.error}</Alert>;
    }

    root.render(
        <>
            <Header showSignIn={false} />
            <Card className="login-container">
                <Card.Header>Sign In</Card.Header>
                <Card.Body>
                    {alert}
                    <Form method="post">
                        <Form.Group className="mb-3">
                            <Form.Label>Username</Form.Label>
                            <Form.Control name="_username" required type="text" />
                        </Form.Group>
                        <Form.Group className="mb-3">
                            <Form.Label>Password</Form.Label>
                            <Form.Control name="_password" required type="password" />
                        </Form.Group>
                        <input name="_csrf_token" type="hidden" value={window.token} />
                        <Button variant="primary" type="submit">
                            Submit
                        </Button>
                    </Form>
                </Card.Body>
            </Card>
        </>,
    );
}
