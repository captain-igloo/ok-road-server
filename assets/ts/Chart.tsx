import {
    BarElement,
    Chart as ChartJS,
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend,
} from 'chart.js';
import * as moment from 'moment';
import * as React from 'react';
import { Bar } from "react-chartjs-2";
import { useSelector } from 'react-redux';

import { RootState } from './store';

ChartJS.register(
    BarElement,
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend,
);

export const options = {
    maintainAspectRatio: false,
    plugins: {
      legend: {
        display: false,
      },
    },
    responsive: true,
  };

export default function Chart() {
    const features = useSelector((state: RootState) => state.okRoad.features);

    const keys = Object.keys(features);
    const values = Array(60).fill(0);
    const labels = [];

    if (keys.length > 0) {
        const diff = features[keys[keys.length - 1]].timestamp - features[keys[0]].timestamp;
        const first = features[keys[0]].timestamp;
        const a = diff / 59;
        if (a > 0) {
            keys.forEach((key) => {
                const timestamp = features[key].timestamp;
                const index = Math.floor((timestamp - first) / a);
                values[index]++;
            });

            for (let i = first; i <= features[keys[keys.length - 1]].timestamp; i += a) {
                labels.push(moment(new Date(i * 1000)).format('h:mm a'));
            }
        }
    }

    return (
        <div className="chart-container">
            <Bar options={options} data={{
                labels,
                datasets: [{
                    backgroundColor: 'rgba(255, 99, 132, 0.5)',
                    data: values,
                }],
            }} />
        </div>
    );
}