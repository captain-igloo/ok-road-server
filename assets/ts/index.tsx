import * as Sentry from '@sentry/browser';
import * as React from 'react';
import Container from 'react-bootstrap/Container';
import * as ReactDOM from 'react-dom/client';
import { Provider } from 'react-redux';

import Header from './Header';
import Important from './Important';
import { setupStore } from './store';

const element = document.getElementById('app');
if (element) {
    const dataParams = element.getAttribute('data-params');
    let configuration;
    if (dataParams) {
        configuration = JSON.parse(dataParams);
        if (configuration.sentryDsn) {
            Sentry.init({
                dsn: configuration.sentryDsn,
            });
        }
    }
    const root = ReactDOM.createRoot(element);
    root.render(
        <Provider store={setupStore()}>
            <Header user={configuration.user} />
            <Container style={{ paddingTop: '5px' }}>
                <h1>Monitor Young Drivers</h1>
                <p>
                    Install an app on their device and keep track of:
                </p>
                <ul>
                    <li>Where they are</li>
                    <li>How fast they are driving</li>
                    <li>Their speed in relation to the speed limit</li>
                </ul>
                <Important>
                    Speed limit data is currently only available in New Zealand.
                </Important>
                <h2>Getting Started</h2>
                <ol>
                    <li>
                        Click
                        {' '}
                        <a href="/register">here</a>
                        {' '}
                        to register with OK Road.  Remember your username and password because you
                        will need it in step 3.
                    </li>
                    <li>
                        Install
                        {' '}
                        <a href="https://owntracks.org">Owntracks</a>
                        {' '}
                        on the device you wish to monitor.
                    </li>
                    <li>
                        Configure Owntracks to publish to OK Road:
                        <ul>
                            <li>Set Preferences / Connection / Endpoint / Host to "192.9.176.238".</li>
                            <li>Set Preferences / Connection / Endpoint / Port to "1883".</li>
                            <li>Set Preferences / Connection / Credentials / Username to your username.</li>
                            <li>Set Preferences / Connection / Credentials / Password to your password.</li>
                        </ul>
                    </li>
                    <li>
                        Make sure that Owntracks has permission to use location services:
                        <ul>
                            <li>Set Settings / Location / Owntracks / Location access to "Allow all the time".</li>
                            <li>Set Settings / Location / Owntracks / Use precise location to true.</li>
                        </ul>
                    </li>
                </ol>
                <img alt="Screenshot" src="/img/okroad.png" />
            </Container>
        </Provider>,
    );
}
