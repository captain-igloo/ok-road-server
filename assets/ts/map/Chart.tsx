import {
    BarElement,
    Chart as ChartJS,
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend,
} from 'chart.js';
import moment from 'moment';
import * as React from 'react';
import { Bar } from 'react-chartjs-2';
import { useSelector } from 'react-redux';

import { RootState } from '../store';

ChartJS.register(
    BarElement,
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend,
);

export const options = {
    maintainAspectRatio: false,
    plugins: {
        legend: {
            display: false,
        },
    },
    responsive: true,
};

const bucketCount = 60;

export default function Chart() {
    const features = useSelector((state: RootState) => state.map.features);
    const fromDate = useSelector((state: RootState) => state.map.fromDate);
    const toDate = useSelector((state: RootState) => state.map.toDate);

    const values = Array(bucketCount).fill(0);
    const labels = [];
    const bucketSize = (toDate - fromDate) / bucketCount;

    if (bucketSize > 0) {
        features.forEach((feature) => {
            const index = Math.floor(((feature.timestamp * 1000) - fromDate) / bucketSize);
            if (index >= 0 && index < bucketCount) {
                values[index] += 1;
            }
        });
        for (let t = fromDate; t < toDate; t += bucketSize) {
            labels.push(moment(new Date(t)).format('h:mm a'));
        }
    }

    return (
        <div className="chart-container">
            <Bar
                options={options}
                data={{
                    labels,
                    datasets: [{
                        backgroundColor: 'rgba(255, 99, 132, 0.5)',
                        data: values,
                    }],
                }}
            />
        </div>
    );
}
