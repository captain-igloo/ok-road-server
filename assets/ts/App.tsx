import * as moment from 'moment';
import * as React from 'react';
import Card from 'react-bootstrap/Card';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import Form from 'react-bootstrap/Form';
import { useDispatch, useSelector } from 'react-redux';

import Chart from './Chart';
import Devices from './Devices';
import Header from './Header';
import Map from './Map';
import { setFromDate, setShowRecent, setToDate } from './slice';
import { AppDispatch, RootState } from './store';

export const useAppDispatch: () => AppDispatch = useDispatch;

export default function App() {
    const bounds = useSelector((state: RootState) => state.okRoad.bounds);
    const fromDate = useSelector((state: RootState) => state.okRoad.fromDate);
    const showRecent = useSelector((state: RootState) => state.okRoad.showRecent);
    const toDate = useSelector((state: RootState) => state.okRoad.toDate);
    
    const dispatch = useAppDispatch();

    return (
        <>
            <Header />
            <Container className="main-container" fluid>
                <Row className="main-container--row">
                    <Col className="left-container" xs={3}>
                        <Card>
                            <Card.Header>Search</Card.Header>
                            <Card.Body>
                                <Form>
                                    <Devices />
                                    <Form.Group className="mb-3">
                                        <Form.Check
                                            checked={showRecent}
                                            id="show-recent"
                                            label="Last 24 hours"
                                            onChange={(e) => {
                                                dispatch(setShowRecent(e.target.checked));
                                            }}
                                            type="checkbox"
                                        />
                                    </Form.Group>
                                    <Form.Group className="mb-3">
                                        <Form.Label>From:</Form.Label>
                                        <Form.Control
                                            disabled={showRecent}
                                            onChange={(e) => {
                                                dispatch(setFromDate((new Date(e.target.value)).getTime()));
                                            }}
                                            type="datetime-local"
                                            value={moment(new Date(fromDate)).format('Y-MM-DDTHH:mm')}
                                        />
                                    </Form.Group>
                                    <Form.Group className="mb-3">
                                        <Form.Label>To:</Form.Label>
                                        <Form.Control
                                            disabled={showRecent}
                                            onChange={(e) => {
                                                dispatch(setToDate((new Date(e.target.value)).getTime()));
                                            }}
                                            type="datetime-local"
                                            value={moment(new Date(toDate)).format('Y-MM-DDTHH:mm')}
                                        />
                                    </Form.Group>
                                    <Form.Group>
                                        <Form.Check label="Fit map" type="checkbox" />
                                    </Form.Group>
                                </Form>
                            </Card.Body>
                        </Card>
                    </Col>
                    <Col className="right-container" xs={9}>
                        <Map bounds={bounds} />
                        <Chart />
                    </Col>
                </Row>
            </Container>
        </>        
    );
}
