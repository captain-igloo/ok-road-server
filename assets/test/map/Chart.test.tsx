import '../setup';

import { describe, expect, test } from '@jest/globals';
import { render } from '@testing-library/react';
import * as React from 'react';
import { Provider } from 'react-redux';

import Chart from '../../ts/map/Chart';
import { setupStore } from '../../ts/store';

describe('Chart component', () => {
    test('Chart renders properly', () => {
        const { container } = render(
            <Provider store={setupStore({
                map: {
                    features: [{
                        id: 1,
                        insertTimestamp: 1744761600,
                        speedLimit: {
                            description: 'WELLINGTON CITY UTA',
                            speedLimit: 50,
                        },
                        timestamp: 1744761600,
                        velocity: 40,
                    }],
                    fromDate: new Date('2025-04-16 00:00:00').getTime(),
                    toDate: new Date('2025-04-17 00:00:00').getTime(),
                } as any,
            })}>
                <Chart />
            </Provider>
        );
        expect(container).toMatchSnapshot();
    });
});
