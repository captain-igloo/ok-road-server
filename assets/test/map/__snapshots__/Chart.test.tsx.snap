// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart component Chart renders properly 1`] = `
<div>
  <div
    class="chart-container"
  >
    <canvas
      height="0"
      role="img"
      style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
      width="0"
    />
  </div>
</div>
`;
